src/client/datascience/kernel-launcher/remoteKernelFinder.ts

```diff
@@ -0,0 +1,160 @@
+import { traceDecorators } from '../../logging';
+import { PythonEnvironment } from '../../pythonEnvironments/info';
+import { captureTelemetry } from '../../telemetry';
+import { Telemetry } from '../constants';
+import { findPreferredKernelIndex } from '../jupyter/kernels/helpers';
+import {
+    KernelConnectionMetadata,
+    DefaultKernelConnectionMetadata,
+    LiveKernelConnectionMetadata
+} from '../jupyter/kernels/types';
+import { PreferredRemoteKernelIdProvider } from '../notebookStorage/preferredRemoteKernelIdProvider';
+import {
+    IJupyterKernelSpec,
+    IJupyterSessionManager,
+    IJupyterSessionManagerFactory,
+    INotebookProviderConnection
+} from '../types';
+import { isInterpreter } from './localKernelFinder';
+import { IRemoteKernelFinder } from './types';
+
+// and finally on the default locations that jupyter installs kernels on.
+@injectable()
+export class RemoteKernelFinder implements IRemoteKernelFinder {
+    /**
+     * List of ids of kernels that should be hidden from the kernel picker.
+     */
+    private readonly kernelIdsToHide = new Set<string>();
+    constructor(
+        @inject(IDisposableRegistry) disposableRegistry: IDisposableRegistry,
+        @inject(PreferredRemoteKernelIdProvider)
+        private readonly preferredRemoteKernelIdProvider: PreferredRemoteKernelIdProvider,
+        @inject(IJupyterSessionManagerFactory) private jupyterSessionManagerFactory: IJupyterSessionManagerFactory
+    ) {
+        disposableRegistry.push(
+            this.jupyterSessionManagerFactory.onRestartSessionCreated(this.addKernelToIgnoreList.bind(this))
+        );
+        disposableRegistry.push(
+            this.jupyterSessionManagerFactory.onRestartSessionUsed(this.removeKernelFromIgnoreList.bind(this))
+        );
+    }
+    @traceDecorators.verbose('Find remote kernel spec')
+    @captureTelemetry(Telemetry.KernelFinderPerf)
+    public async findKernel(
+        resource: Resource,
+        connInfo: INotebookProviderConnection | undefined,
+        option?: nbformat.INotebookMetadata | PythonEnvironment,
+        _cancelToken?: CancellationToken
+    ): Promise<KernelConnectionMetadata | undefined> {
+        // Get list of all of the specs
+        const kernels = await this.listKernels(resource, connInfo);
+
+        // Find the preferred kernel index from the list.
+        const notebookMetadata = option && !isInterpreter(option) ? option : undefined;
+        const preferred = findPreferredKernelIndex(
+            kernels,
+            resource,
+            [],
+            notebookMetadata,
+            undefined,
+            this.preferredRemoteKernelIdProvider
+        );
+        if (preferred >= 0) {
+            return kernels[preferred];
+        }
+    }
+
+    // Talk to the remote server to determine sessions
+    @captureTelemetry(Telemetry.KernelListingPerf)
+    public async listKernels(
+        _resource: Resource,
+        connInfo: INotebookProviderConnection | undefined
```

Comment:
Would be good to accept a cancellation token here, plenty of promises inside this.